raphtory/src/search/graph_index.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{
+    core::{
+        entities::{EID, VID},
+        storage::timeindex::{AsTime, TimeIndexEntry},
+        utils::errors::GraphError,
+    },
+    db::api::storage::graph::storage_ops::GraphStorage,
+    prelude::*,
+    search::{edge_index::EdgeIndex, node_index::NodeIndex},
+};
+use std::fmt::{Debug, Formatter};
+use tantivy::query::QueryParser;
+
+#[derive(Clone)]
+pub struct GraphIndex {
+    pub(crate) node_index: NodeIndex,
+    pub(crate) edge_index: EdgeIndex,
+}
+
+impl Debug for GraphIndex {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GraphIndex")
+            .field("node_index", &self.node_index)
+            .field("edge_index", &self.edge_index)
+            .finish()
+    }
+}
+
+impl<'a> TryFrom<&'a GraphStorage> for GraphIndex {
+    type Error = GraphError;
+
+    fn try_from(graph: &GraphStorage) -> Result<Self, Self::Error> {
+        let node_index = NodeIndex::index_nodes(graph)?;
+        // node_index.print()?;
+
+        let edge_index = EdgeIndex::index_edges(graph)?;
+        // edge_index.print()?;
+
+        Ok(GraphIndex {
+            node_index,
+            edge_index,
+        })
+    }
+}
+
+impl GraphIndex {
+    pub fn new() -> Self {
+        GraphIndex {
+            node_index: NodeIndex::new(),
+            edge_index: EdgeIndex::new(),
+        }
+    }
+
+    pub fn print(&self) -> Result<(), GraphError> {
+        self.node_index.print()?;
+        self.edge_index.print()?;
+        Ok(())
+    }
+
+    pub(crate) fn node_parser(&self) -> Result<QueryParser, GraphError> {
+        Ok(QueryParser::for_index(&self.node_index.index, vec![]))
+    }
+
+    pub(crate) fn edge_parser(&self) -> Result<QueryParser, GraphError> {
+        Ok(QueryParser::for_index(&self.edge_index.index, vec![]))
+    }
+
+    pub(crate) fn add_node_update(
+        &self,
+        graph: &GraphStorage,
+        t: TimeIndexEntry,
+        v: VID,
+    ) -> Result<(), GraphError> {
+        self.node_index.add_node_update(graph, t, v)
+    }
+
+    pub(crate) fn add_edge_update(
+        &self,
+        graph: &GraphStorage,
+        edge_id: EID,
+        t: TimeIndexEntry,
+        src: VID,
+        dst: VID,
+        layer: usize,
+        props: &[(usize, Prop)]
+    ) -> Result<(), GraphError> {
+        self.edge_index
+            .add_edge_update(graph, edge_id, t, src, dst, layer, props)
+    }
```

Comment:
need separate methods for constant property updates for nodes and edges and likely also methods for creating new nodes and edges so you don't have to worry about duplicate documents in the index.

Reply:
Make sure to handle overwriting of constant property values correctly (the old value no longer exists and shouldn't show up in search results)!